File: app-ios-nativa/Fontaneria/Fuentes/Fontaneria/Sincronizador.swift
import Dominio
import Foundation

/// Un ciclo completo de sincronización: primero sube lo que hay pendiente,
/// después baja lo que haya cambiado en el servidor.
///
/// Cómo fusionar lo decide `Dominio`; aquí solo se encadenan almacén, cliente
/// y sesión.
///
/// Vive en el hilo principal porque comparte el almacén con la interfaz, que
/// lo lee para pintar la lista. Las esperas de red son asíncronas, así que no
/// bloquean nada: lo único que ocurre en el hilo principal son las consultas
/// a SQLite, que tardan microsegundos.
@MainActor
public final class Sincronizador {
    private let almacen: AlmacenLocal
    private let cliente: ClienteApi
    private let sesion: Sesion

    public init(almacen: AlmacenLocal, cliente: ClienteApi, sesion: Sesion) {
        self.almacen = almacen
        self.cliente = cliente
        self.sesion = sesion
    }

    /// Devuelve cuántos cambios locales rechazó el servidor (normalmente cero).
    @discardableResult
    public func sincronizar() async throws -> Int {
        let rechazados = try await subirPendientes()
        try await bajarCambios()
        return rechazados
    }

    private func subirPendientes() async throws -> Int {
        let elementos = Array(try almacen.cargarPendientes().values)
        let etiquetas = Array(try almacen.cargarEtiquetasPendientes().values)
        guard !elementos.isEmpty || !etiquetas.isEmpty else {
            return 0
        }

        let respuesta = try await sesion.conReintento { token in
            try await cliente.subirCambios(
                elementos: elementos,
                etiquetas: etiquetas,
                tokenAcceso: token
            )
        }

        try almacen.guardar(
            Sincronizacion.aplicarRespuestaPush(
                try almacen.cargarTodos(),
                definitivos: respuesta.elementos
            )
        )
        try almacen.guardarEtiquetasDefinidas(
            Sincronizacion.aplicarRespuestaPushEtiquetas(
                try almacen.cargarEtiquetasDefinidas(),
                definitivas: respuesta.etiquetasDefinidas
            )
        )

        // Lo rechazado sale de la cola igual que lo aceptado. El servidor no
        // lo va a admitir por mucho que se insista, y dejarlo dentro reenvía
        // el lote entero en cada sincronización, para siempre y sin que se
        // note: así se quedaron atascados los enlaces importados a una cuenta
        // nueva.
        try almacen.limpiarPendientes(
            respuesta.elementos.map(\.id) + respuesta.rechazados.map(\.id)
        )
        try almacen.limpiarEtiquetasPendientes(
            respuesta.etiquetasDefinidas.map(\.id) + respuesta.etiquetasRechazadas.map(\.id)
        )

        return respuesta.rechazados.count + respuesta.etiquetasRechazadas.count
    }

    private func bajarCambios() async throws {
        // Se repite mientras el servidor diga que quedan páginas: biblioteca
        // grande o primera sincronización.
        while true {
            let desde = try almacen.cursor()
            let respuesta = try await sesion.conReintento { token in
                try await cliente.bajarCambios(desde: desde, tokenAcceso: token)
            }

            try almacen.guardar(
                Sincronizacion.aplicarPull(
                    try almacen.cargarTodos(),
                    recibidos: respuesta.elementos,
                    pendientes: try almacen.cargarPendientes()
                )
            )
            try almacen.guardarEtiquetasDefinidas(
                Sincronizacion.aplicarPullEtiquetas(
                    try almacen.cargarEtiquetasDefinidas(),
                    recibidas: respuesta.etiquetasDefinidas,
                    pendientes: try almacen.cargarEtiquetasPendientes()
                )
            )

            if respuesta.masDisponible, !respuesta.elementos.isEmpty {
                try almacen.fijarCursor(respuesta.elementos.map(\.actualizadoEn).max() ?? 0)
            } else {
                try almacen.fijarCursor(respuesta.servidorEn)
                return
            }
        }
    }
}

File: app-android/fontaneria/src/main/kotlin/com/jmortizsilva/guardarenlaces/fontaneria/Sincronizador.kt
package com.jmortizsilva.guardarenlaces.fontaneria

import com.jmortizsilva.guardarenlaces.dominio.Sincronizacion

/**
 * Un ciclo completo de sincronización: primero sube lo que hay pendiente, después baja lo que haya
 * cambiado en el servidor.
 *
 * Cómo fusionar lo decide `dominio`; aquí solo se encadenan almacén, cliente y sesión. Cada fusión
 * se guarda en una transacción: leer, fusionar y escribir de golpe, para que un cambio que haga la
 * pantalla entretanto no se quede a medias entre una lectura y una escritura.
 */
class Sincronizador(
    private val almacen: AlmacenLocal,
    private val cliente: ClienteApi,
    private val sesion: Sesion,
) {
    /** Devuelve cuántos cambios locales rechazó el servidor (normalmente cero). */
    suspend fun sincronizar(): Int {
        val rechazados = subirPendientes()
        bajarCambios()
        return rechazados
    }

    private suspend fun subirPendientes(): Int {
        val elementos = almacen.cargarPendientes().values.toList()
        val etiquetas = almacen.cargarEtiquetasPendientes().values.toList()
        if (elementos.isEmpty() && etiquetas.isEmpty()) return 0

        val respuesta = sesion.conReintento { token ->
            cliente.subirCambios(elementos, etiquetas, token)
        }

        almacen.enTransaccion {
            almacen.guardar(
                Sincronizacion.aplicarRespuestaPush(almacen.cargarTodos(), respuesta.elementos)
            )
            almacen.guardarEtiquetasDefinidas(
                Sincronizacion.aplicarRespuestaPushEtiquetas(
                    almacen.cargarEtiquetasDefinidas(),
                    respuesta.etiquetasDefinidas,
                )
            )
            // Lo rechazado sale de la cola igual que lo aceptado. El servidor no lo va a admitir
            // por mucho que se insista, y dejarlo dentro reenvía el lote entero en cada
            // sincronización, para siempre y sin que se note: así se quedaron atascados en iOS
            // los enlaces importados a una cuenta nueva.
            almacen.limpiarPendientes(
                respuesta.elementos.map { it.id } + respuesta.rechazados.map { it.id }
            )
            almacen.limpiarEtiquetasPendientes(
                respuesta.etiquetasDefinidas.map { it.id } +
                    respuesta.etiquetasRechazadas.map { it.id }
            )
        }
        return respuesta.rechazados.size + respuesta.etiquetasRechazadas.size
    }

    private suspend fun bajarCambios() {
        // Se repite mientras el servidor diga que quedan páginas: biblioteca grande o primera
        // sincronización.
        while (true) {
            val desde = almacen.cursor()
            val respuesta = sesion.conReintento { token -> cliente.bajarCambios(desde, token) }

            val terminado = !respuesta.masDisponible || respuesta.elementos.isEmpty()
            almacen.enTransaccion {
                almacen.guardar(
                    Sincronizacion.aplicarPull(
                        almacen.cargarTodos(),
                        respuesta.elementos,
                        almacen.cargarPendientes(),
                    )
                )
                almacen.guardarEtiquetasDefinidas(
                    Sincronizacion.aplicarPullEtiquetas(
                        almacen.cargarEtiquetasDefinidas(),
                        respuesta.etiquetasDefinidas,
                        almacen.cargarEtiquetasPendientes(),
                    )
                )
                // El cursor, en la misma transacción que lo que ha bajado: si la app se cierra a
                // mitad, o se guardan las dos cosas o ninguna, y la próxima vez se pide lo mismo.
                almacen.fijarCursor(
                    if (terminado) respuesta.servidorEn
                    else respuesta.elementos.maxOf { it.actualizadoEn }
                )
            }
            if (terminado) return
        }
    }
}
